src/components/Login.tsx

```diff
@@ -1,19 +1,82 @@
 import type { Theme } from "@emotion/react";
 import { useTheme } from "@emotion/react";
 import { css } from "@emotion/react";
+import { useRef, useState } from "react";
 import { useNavigate } from "react-router";
+import Input from "@/components/Input";
 
 const Login = () => {
   const theme = useTheme();
   const navigate = useNavigate();
+  const ref = useRef(null);
+
+  const validateEmail = (email: string) => {
+    const EMAIL_REGEXP = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
+    return Boolean(EMAIL_REGEXP.test(email));
+  };
+
+  const validatePassword = (password: string) => {
+    return password.length >= 8;
+  };
+
+  function useValidate(
+    validator: (value: string) => boolean,
+    errorMessage: string
+  ) {
+    const [string, setString] = useState("");
+    const [isValid, setIsValid] = useState(false);
+    const [message, setMessage] = useState("");
+
+    const onChange = (e: any) => {
```

Comment:
```suggestion
    const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
```

any와 멀어질수록.. TS와는 가까워집니다!
가급적 사용하지 않도록 해보셔요. 

정모르겠으면 AI 한테 던져도 되고요, 그래도 이상한 답변을 내놓는다면 저에게 질문 주셔요.

Reply:
[Change Request] 2. setState 버그 수정

여기는 또 그냥 넘어갈 수 없는 포인트가 하나 있어서 수정 요청 드립니다!

리액트에서 setState를 하면 상태가 바로 반영되지 않는 거 알고 계셨나요..?
리액트의 setState는 `비동기함수`입니다.

참고 : https://ko.react.dev/learn/choosing-the-state-structure#avoid-redundant-state

완전 FE 면접 단골 질문이에요.

비동기라는 개념을 아직 안 배우셨다면, 코드 순서대로 실행되는 코드가 아니다라고만 일단 이해해보시면 좋을 거 같아요.
그럼 언제 실행 되는가? -> 리액트 컴포넌트가 브라우저에 다시 그려질 때 실행됩니다.

그래서 지금 유효한 email 을 넣었음에도 버튼이 비활성화 된 걸 볼 수있죠??
valid 판단을 `1233@a.` 스트링으로 해서 그렇습니다.

setString 한 값이 아직 반영이 안 된 거죠 ㅠ

[setState의 비동기 동작으로 인한 오류 캡처]

<img width="640" alt="image" src="https://github.com/user-attachments/assets/4ae88d37-dc50-4a1e-8dff-52c0c89447af" />

그럼 어떻게 하면 될까요??

```ts
const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
  const nextValue = e.target.value;

  setString(nextValue);
  setIsValid(validator(nextValue));
};
```

요런 식으로 바꾸시면 됩니다.
setString을 한 state를 참조하는게 아니라 inputElement의 이벤트에서 넘어온 value를 참조하는 거죠.

똑같아 보이지만 내부적으로 엄청난 차이가 있습니다..!!
완전 중요한 개념이니 꼭 변경 적용 해보시면서 머릿 속에 오래 남아있었으면 좋겠습니다.

Reply:
엄청 중요한 부분이었네요. 공식문서 reference 통해서 추가학습 해보겠습니다!